fast64_internal/f3d/f3d_material.py

```diff
@@ -1722,39 +1722,47 @@ def update_tex_values_manual(material: bpy.types.Material, context, prop_path=No
     set_texture_settings_node(material)
 
 
-def getMaterialScrollDimensions(material):
-    useDict = all_combiner_uses(material)
-
-    if (
-        useDict["Texture 0"]
-        and material.tex0.tex is not None
-        and useDict["Texture 1"]
-        and material.tex1.tex is not None
-        and material.tex0.tex.size[0] > 0
-        and material.tex0.tex.size[1] > 0
-        and material.tex1.tex.size[0] > 0
-        and material.tex1.tex.size[1] > 0
-    ):
-        if material.uv_basis == "TEXEL0":
-            return material.tex0.tex.size
-        else:
-            return material.tex1.tex.size
-    elif (
-        useDict["Texture 1"]
-        and material.tex1.tex is not None
-        and material.tex1.tex.size[0] > 0
-        and material.tex1.tex.size[1] > 0
-    ):
-        return material.tex1.tex.size
-    elif (
-        useDict["Texture 0"]
-        and material.tex0.tex is not None
-        and material.tex0.tex.size[0] > 0
-        and material.tex0.tex.size[1] > 0
-    ):
-        return material.tex0.tex.size
+def shift_num(num: int, amt: int):
+    if amt < 0:
+        return num >> -amt
+    return num << amt
+
+
+def shift_dimensions(tex_prop: "TextureProperty", dimensions: tuple[int, int]):
+    shifted = (shift_num(dimensions[0], tex_prop.S.shift), shift_num(dimensions[1], tex_prop.T.shift))
+    s_mirror_scale = 2 if tex_prop.S.mirror else 1
+    t_mirror_scale = 2 if tex_prop.T.mirror else 1
+    return (shifted[0] * s_mirror_scale, shifted[1] * t_mirror_scale)
+
+
+def getMaterialScrollDimensions(f3dMat):
+    texDimensions0 = None
+    texDimensions1 = None
+    useDict = all_combiner_uses(f3dMat)
+
+    if useDict["Texture 0"] and f3dMat.tex0.tex_set:
+        if f3dMat.tex0.use_tex_reference:
+            texDimensions0 = f3dMat.tex0.tex_reference_size
+        elif f3dMat.tex0.tex:
+            texDimensions0 = (f3dMat.tex0.tex.size[0], f3dMat.tex0.tex.size[1])
+
+    if useDict["Texture 1"] and f3dMat.tex1.tex_set:
+        if f3dMat.tex1.use_tex_reference:
+            texDimensions1 = f3dMat.tex1.tex_reference_size
+        elif f3dMat.tex0.tex:
+            texDimensions1 = (f3dMat.tex1.tex.size[0], f3dMat.tex1.tex.size[1])
+
+    if texDimensions0 is not None:
+        texDimensions0 = shift_dimensions(f3dMat.tex0, texDimensions0)
     else:
-        return [32, 32]
+        texDimensions0 = (4, 4)
```

Comment:
muh 2x2 or 8x1 textures sadge

Reply:
its a fail safe if something is undefined 😔 plus the other texture's value will override this since its the max between the two

Reply:
Imo would be better as 0,0 and a min(..., ...) in the return?

But really textures can be 1px high afaik
And 2px wide if rgba32 (so the tex line size is 8-aligned, iirc that's the only requirement)

Reply:
> Imo would be better as 0,0 and a min(..., ...) in the return?

why min? this specifically needs the max width/height between the two textures. 

> But really textures can be 1px high afaik
> And 2px wide if rgba32
then `(2, 1)` seems like a sane default min yeah?

Reply:
About the min:
Me stupid, I meant max
Like `max(1, dim0, dim1)`

And why not just 1 for both s and t regardless of what's possible, it's not like it matters at that point anyway

Reply:
i guess im just being extra safe but idc either way. i need to get this fix in so i'll just do 1, 1

Reply:
(also i just realized that `max` allows *args 💀)

Reply:
okay @Dragorn421 pushed and updated